### ISSUE TITLE
Problems with variable names when consulting DCG and listing/1

### ISSUE THREAD

**Issue by jburse** [SWI-Prolog/swipl-devel#313]
I was taking hands on some old code that I had written in 2012.

I have some DCG which I am consulting from file. I am using new
SWI-Prolog 7. It seems that listing/1 doesn't show the clauses correctly.
This is the DCG code:

```
% thm(+Sequent, +Formula)
:- dynamic thm/4.
thm(S, A) -->
   {member(A, S)}, [id].
thm(S, (A->B)) -->
   thm([A|S], B), [right].
thm(S, C) -->
   {select((A->B), S, T)},
   thm(S, A),
   thm([B|T], C), [left].
```

Somehow my DCG code didn't work. So I started listing it. And this is the 
result, which is either wrong or either only wrongly displayed. There
are also problems with the other clauses:

![grafik](https://user-images.githubusercontent.com/5411487/42179112-14f9d0d0-7e33-11e8-903b-688ee3966da3.png)

**Comment by JanWielemaker** [comment]
This doesn't seem a complete program.  DCG doesn't produce multiple clauses from a single grammar rule.

**Comment by jburse** [comment]
Its 3 DCG input clauses. I thought showing only the first is enough. 
But I made an edit you can see all of them, which I already 
created in 2012. 

It happens that they use a variable "A" and/or a variable "B"
either in the head or in the body. Which then gives a wrong
result when using listing/1.

I had the same problem in the past in my system. The cure is
that you need to adapt the strategy of generating names unnamed
variables, singleton or not, based on the named variables.

DCG introduces two or more or less variables in clause through
its translation, which are not named in the input source. The predicate
that does this name invention for me is:

sys_number_variables(V, N, S, M):
    The predicate succeeds with variable names M resulting from giving names to the variables in V, respecting the variable names N and the unnamed singletons S.

**Comment by jburse** [comment]
When using sys_number_variables/4 I get. Also some (=)/2 are
optimized away, but this is not the content of the raised issue:

![grafik](https://user-images.githubusercontent.com/5411487/42181477-3ef91262-7e3b-11e8-942d-3bd325a9801d.png)

**Comment by JanWielemaker** [comment]
Git it.  Should be fixed with 9b7ef8243050fc430cb3ec86ca70c6ad6b93e433
